### PR TITLE
Improve wrapper features

### PR DESCRIPTION
**Pull request by roigcarlo**

### CODE REVIEW

**Comment by roigcarlo** [comment]
@nathanfranklin It looks like automatic Kratos releases are compiled with python3, which is making kratos to load incorrectly. I'm currently compiling a version with python 2.7.4, but it may take a while.

File: use_cases/CFD_example.gid/test_kratosWrapper.py [hunk 1]
```diff
@@ -39,7 +49,7 @@
 CFDWrapper.setMeshData(mesh)
 CFDWrapper.add_mesh(mesh)
 
-for i in xrange(0,10):
+for i in xrange(0,CFDWrapper.CM[CUBA.NUMBER_OF_TIME_STEPS]):
```
Comment (nathanfranklin) [inline use_cases/CFD_example.gid/test_kratosWrapper.py:52]: I think the '''CFDCFDWrapper.CM[CUBA.NUMBER_OF_TIME_STEPS]''' should
be describing the number of time steps that occur when '''run''' occurs.  

For example:

```
CFDWrapper.CM[CUBA.NUMBER_OF_TIME_STEPS] = 5
for i in xrange(0, 10):
  CFDWrapper.run()  # runs 5 steps

#has now run a total of 50 steps
```

**Comment by roigcarlo** [comment]
hi @nathanfranklin,

I fixed your comment and kratos is ( finally ) running on travis.

We still have some problems with the DEM tests from our side ( it runs correctly but the results are not the expected due to a difference between the test case version and kratos version ). As soon as it gets solved i will ping you.

File: wrappers/kratosWrapper.py [hunk 2]
```diff
@@ -51,68 +47,127 @@ def __init__(self):
...
 
-    def __addNodalVariablesToModelpart(self):
-        pass
+    # ABCModelingEngine Implementation
+
+    def add_particles(self, particle_container):
+        message = 'KratosWrapper does not handle particle container'
+        raise NotImplementedError(message)
+
+    def add_mesh(self, src):
```
Comment (nathanfranklin) [inline wrappers/kratosWrapper.py]: This deviates from the ABCModelingEngine as:
1) it does not return anything
2) just keeps the original 'src' mesh.  

For example, (2) would be a problem:

```
 with closing(H5CUDS.open("foo.cuds", 'r')) as file_handle:   
        wrapper.add_mesh(file_handle.get_mesh("meshA"))
        wrapper.run()
```

something along the lines of this should be done:

```
def add_mesh(self, src):
    copy_mesh = Mesh(name=src.name)
    copy_mesh.data = src.data
    copy_mesh.data = H5Mesh(group, self._handle)
    for point in src.iter_points():
         copy_mesh.add_point(point)

    ...
    ...
    self.meshes[src.name] = copy_mesh
    return copy_mesh
```

Comment (nathanfranklin) [inline wrappers/kratosWrapper.py]: Note, my suggestion of what to be done is also faulty as it wouldn't handle renaming of the mesh (`mesh.name="new_name"`) as the wrapper would be unaware of such changes:

**Comment by roigcarlo** [comment]
@nathanfranklin Thx for the feedback. I see that add_mesh was not behaving as expected. I have changed the way the wrapper handles the meshes so it acts according to the description.
I have also fixed the problems with travis ( again, hope it doesn't crash anymore) so integration should start to appear in green.

I am still changing the iter_mesh functions as I see that it does not support iterating over subsets. 
As soon as I have it I will ping you.

**Comment by itziakos** [comment]
@roigcarlo, you might want to use the mesh check templates when testing the Mesh behaviour.
you can see example usage:
- https://github.com/simphony/simphony-common/blob/master/simphony/io/tests/test_h5_mesh.py
- https://github.com/simphony/simphony-mayavi/pull/66/files#diff-93100555da9d1e915df12a581b5f457eR1

**Comment by roigcarlo** [comment]
@itziakos I'll take a look thx!

**Comment by itziakos** [comment]
for an engine creating a mesh whould probably be:

``` python
def setUp(self):
      self.engine = Engine()
      # .... execute the parent setUp()

def create_container(self, name):
      mesh = Mesh(name=name)
      return self.engine.add_mesh(mesh)
```

**Comment by roigcarlo** [comment]
@itziakos I've been through the examples you sent me but I cannot see how this is applicable to my wrapper.

Basically the wrapper uses a standard cuds mesh. When the run function is called  the mesh is converted to Kratos format at once and when it has finished to run is converted back to cuds. ( see: . 

https://github.com/simphony/simphony-kratos/blob/improve-wrapper-features/wrappers/kratos_CFD_wrapper.py#L569-L581 

and

https://github.com/simphony/simphony-kratos/blob/improve-wrapper-features/wrappers/kratos_CFD_wrapper.py#L629-L641

Basically what the wrapper exposes is a simphony mesh.

The only way in which I see these mesh test having sense here is, if this is how its supposed to be, to make an intermediate "kratos_mesh.py" implementing the ABCMesh which interacts directly with the kratos data structures element by element rather than converting the mesh at the execution.

**Comment by itziakos** [comment]
> Basically the wrapper uses a standard cuds mesh. When the run function is called the mesh is converted to Kratos format at once and when it has finished to run is converted back to cuds. 

I see, the wrapper architecture is similar to a File-IO wrapper but with direct access instead of IO. sorry I missed that. 

No worries then.

File: wrappers/tests/dem/test_kratos_dem.py [hunk 3]
```diff
@@ -0,0 +1,67 @@
+""" test_kratosWrapper module
+
+This module contains the unitary tests for the
+kratosWrapper class.
+
+"""
+import sys
+import os
+
+sys.path.append(os.path.abspath('wrappers/tests/dem'))
```
Comment (nathanfranklin) [inline wrappers/tests/dem/test_kratos_dem.py]: This causes problem when trying to run the tests outside of `simphony-kratos`.  I will open up an issue to follow this.

File: wrappers/tests/dem/test_kratos_dem.py [hunk 4]
```diff
@@ -0,0 +1,67 @@
...
+
+from KratosMultiphysics import *
+from KratosMultiphysics.DEMApplication import *
+from KratosMultiphysics.IncompressibleFluidApplication import *
+from KratosMultiphysics.FluidDynamicsApplication import *
+from KratosMultiphysics.ExternalSolversApplication import *
+from KratosMultiphysics.MeshingApplication import *
```
Comment (nathanfranklin) [inline wrappers/tests/dem/test_kratos_dem.py]: are these includes from KratosMultiphysics needed?

File: wrappers/tests/dem/test_kratos_dem.py [hunk 5]
```diff
@@ -0,0 +1,67 @@
...
+            self.rigid_path
+        )
+
+        wrapper.setMeshData(mesh)
+        wrapper.add_mesh(mesh)
+
+        for i in xrange(0, 1):
+            wrapper.run()
+
+        wrapper.finalize()
```
Comment (nathanfranklin) [inline wrappers/tests/dem/test_kratos_dem.py]: ABCModelingEngine does not have a`finalize` method.

would it be possible to add whatever it does in the `run()` method?

File: wrappers/tests/dem/test_kratos_dem.py [hunk 6]
```diff
@@ -0,0 +1,67 @@
...
+
+        wrapper.CM[CUBA.TIME_STEP] = self.time_step
+        wrapper.CM[CUBA.NUMBER_OF_TIME_STEPS] = self.num_steps
+
+        mesh = wrapper.read_modelpart(
+            self.fluid_path,
+            self.rigid_path
+        )
+
+        wrapper.setMeshData(mesh)
```
Comment (nathanfranklin) [inline wrappers/tests/dem/test_kratos_dem.py]: ABCModelingEngine does not have a `setMeshData` method.

Would it be possible to rename `setMeshData` to `_setMeshData()`
and then have add_mesh() call that first?

File: wrappers/kratos_CFD_wrapper.py [hunk 7]
```diff
@@ -535,46 +561,23 @@ def initialize(self):
...
-
-        # model_part_io_fluid = ModelPartIO("CFD_exampleFluid")
-        # model_part_io_fluid.ReadModelPart(self.fluid_model_part)
-
         # Import the into Kratos
-        self.__importKratosElements(
+        self.importKratosNodes(
             self.get_mesh("Model"),
-            newFluidMp
+            self.fluid_model_part
```
Comment (nathanfranklin) [inline wrappers/kratos_CFD_wrapper.py]: only a mesh named "Model" is ever used?    It appears that if a user adds a mesh with another name (using `add_mesh`), then that mesh is never used.

File: wrappers/tests/dem/test_kratos_dem.py (hunk 3 above)
Comment (nathanfranklin) [inline wrappers/tests/dem/test_kratos_dem.py]: Issue added #14

File: wrappers/kratos_CFD_wrapper.py [hunk 8]
```diff
@@ -400,18 +391,42 @@ def __importKratosConditions(self, src, dst, entitylist=None):
...
+        #     if point.data[CUBAExtension.IMPOSED_PRES] == 1:
+        #         node.Fix(PRESSURE)
+
+        pass
 
     # FileIO
 
     def read_modelpart(self, filename):
-        """ Reads a Kratos formated modelpart NYI
+        """ Reads a Kratos formated modelpart
```
Comment (nathanfranklin) [inline wrappers/kratos_CFD_wrapper.py]: I would have expected that this method reads in a KRATOS file and returns a SimPhoNy mesh (with name "Model") .  is that its intent?

just wondering about lines 435-437 below:

```
        self.fluid_model_part = ModelPart("FluidPart")

        self.addNodalVariablesToModelpart(self.fluid_model_part)
```

Comment (nathanfranklin) [inline wrappers/kratos_CFD_wrapper.py]: Currently, the user needs to an `read_modelpart` where an existing Kratos ModelPart file is used to create a SimPhoNy mesh that can then be added to the wrapper.

Is there a description anywhere (or is it being planned) that describes the SimPhonyMesh that this wrapper needs.

Maybe an issue could be opened to add documentation about what the Mesh should be like (what CUBA should be used on cells, points etc..) for this kratos wrapper.

File: wrappers/kratos_CFD_wrapper.py [hunk 9]
```diff
@@ -615,37 +618,33 @@ def run(self):
         self.fluid_model_part.CloneTimeStep(self.time)
         self.time = self.time + Dt
 
-        for n in xrange(0, substeps):
+        for n in xrange(0, self.CM[CUBA.NUMBER_OF_TIME_STEPS]):
```
Comment (nathanfranklin) [inline wrappers/kratos_CFD_wrapper.py]: I like this here....  but it looks like test_kratosWrapper.py (
here https://github.com/simphony/simphony-kratos/pull/9/files#r32045414 ) has not been updated yet

File: wrappers/tests/cfd/test_kratos_cfd.py [hunk 10]
```diff
@@ -0,0 +1,75 @@
...
+        wrapper.CM[CUBA.NUMBER_OF_TIME_STEPS] = self.num_steps
+        wrapper.BC[CUBA.VELOCITY] = self.bc_vel
+        wrapper.BC[CUBA.PRESSURE] = self.bc_pre
+
+        mesh = wrapper.read_modelpart(self.path)
+
+        wrapper.setMeshData(mesh)
+        wrapper.add_mesh(mesh)
+
+        for i in xrange(0, wrapper.CM[CUBA.NUMBER_OF_TIME_STEPS]):
```
Comment (nathanfranklin) [inline wrappers/tests/cfd/test_kratos_cfd.py]: same as https://github.com/simphony/simphony-kratos/pull/9/files#r32045414

**Comment by nathanfranklin** [comment]
@roigcarlo ,  I added some additional comments.  I don't know how best to proceed.  Maybe, take a look at the comments to see if they make sense first.  Then, possibly, fix the easier ones and for ones that will require more time, maybe it would best that individual issues are made.

A lot of the comments which were made for the DEM wrapper apply to the CFD wrapper, and vice versa.

**Comment by nathanfranklin** [comment]
@roigcarlo ,  i would also suggest prioritizing #11

File: wrappers/tests/dem/test_kratos_dem.py (hunk 4 above)
Comment (roigcarlo) [inline wrappers/tests/dem/test_kratos_dem.py]: Yes. I can try to delete some of them if it is too problematic.

File: wrappers/tests/dem/test_kratos_dem.py (hunk 5 above)
Comment (roigcarlo) [inline wrappers/tests/dem/test_kratos_dem.py]: Yes, It can be done

File: wrappers/tests/dem/test_kratos_dem.py (hunk 6 above)
Comment (roigcarlo) [inline wrappers/tests/dem/test_kratos_dem.py]: This also can be done by the wrapper internally

File: wrappers/kratos_CFD_wrapper.py (hunk 7 above)
Comment (roigcarlo) [inline wrappers/kratos_CFD_wrapper.py]: Thx for ponting this out. This is really wrong. The problem is that you only can run a fixed number of meshes for each wrapper but of course it does not need to be called "Model".

Would be correct to implement something in the lines of

```
SP[CUBAExt.FLUID_MESH_TO_RUN] = "foo"
```

**Comment by roigcarlo** [comment]
@nathanfranklin Thx for the comments. I apologize for not answering yesterday but it was holyday here and I was not checking my mails. I will start uploading the changes as soon as they pass the tests.

**Comment by nathanfranklin** [comment]
Thanks.  Keep me posted.  To include Kratos in the next version of the simphony-framework (so that it can be reported on in D1.5), this PR needs to be merged and and issue #11 and #15 need to be addressed as soon as possible (preferably by Friday). 

if there are any comments for this PR that are difficult to address in this time-frame, then maybe separate issues should be open for them.

**Comment by roigcarlo** [comment]
@nathanfranklin I just updated the README.rst.

I'm about to upload a bunch of changes that will crash travis since I have to move to another machine. this hopefully will help to fix #15 by Friday as requested but the code is still a little bit messy

I will ping you as soon as it works again

File: setup.py [hunk 11]
```diff
@@ -22,17 +22,30 @@ def write_version_py(filename=None):
...
     long_description=README_TEXT,
     entry_points={
-        'simphony.engine': ['kratos = simkratos']},
+        'simphony.engine': [
+            'kratosDEM = wrappers.DEM',
+            'kratosCFD = wrappers.CFD'
+        ]},
     packages=find_packages(),
-    install_requires=["simphony >= 0.1.1"]
+    install_requires=["simphony >= 0.1.5"]
```
Comment (nathanfranklin) [inline setup.py]: could you move this back to >= 0.1.3

(0.1.3 is the goal for integration and I suspect this code runs with 0.1.3, 0.1.4 and 0.1.5)

File: setup.py [hunk 12]
```diff
@@ -22,17 +22,30 @@ def write_version_py(filename=None):
...
+dem = os.path.join(
+    os.path.dirname(__file__),
+    'wrappers/DEM', 'version.py'
+)
+
+write_version_py(cfd)
+write_version_py(dem)
 
 setup(
-    name='simkratos',
```
Comment (itziakos) [inline setup.py]: I would suggest that you use a more specific name for the kratos plugin (e.g. simphony_kratos). This name is used for the package folder inside the `site-package` and using `wrappers` is very generic and could conflict with other packages.

File: setup.py [hunk 13]
```diff
@@ -22,17 +22,30 @@ def write_version_py(filename=None):
...
+    name='wrappers',
```
Comment (itziakos) [inline setup.py]: I would suggest that you use a more specific name for the kratos plugin (e.g. `simphony_kratos`). This name is used for the package folder inside the `site-package` and using `wrappers` is very generic and could conflict with other packages.

File: setup.py [hunk 14]
```diff
@@ -22,17 +22,30 @@ def write_version_py(filename=None):
         fh.close()
 
 
-write_version_py()
+cfd = os.path.join(
```
Comment (itziakos) [inline setup.py]: You probably do not need to do this. The `version.py` is necessary for the package and it written during setup so that the version of the package is controlled by the setup.py.

File: setup.py [hunk 15]
```diff
@@ -5,7 +5,7 @@
 with open('README.rst', 'r') as readme:
     README_TEXT = readme.read()
 
-VERSION = '0.1.0.dev0'
+VERSION = '0.1.5'
```
Comment (itziakos) [inline setup.py]: this is the version of the kratos simphony plugin correct?

File: README.rst [hunk 16]
```diff
@@ -15,11 +15,56 @@ simphony-kratos is hosted on github: https://github.com/simphony/simphony-kratos
...
+    export PYTHONPATH=$PYTHONPATH:$EXTRACT_DIR/kratos
+    
+Usage
+-----
+
+To use any of the wrappers after the installation they need to be imported as plugins.
+
+For Kratos-CFD::
+
+  from simphony.engine import kratos
```
Comment (itziakos) [inline README.rst:58]: Please update the documentation to reflect the plugin entry points as they are setup by `setup.py`

File: setup.py (hunk 13 above)
Comment (itziakos) [inline setup.py]: Please also note that for consistency the main folder will also have to be renamed (e.g. `wrappers -> simphony_kratos`) and the local imports updated as necessary.

File: .travis.yml [hunk 17]
```diff
@@ -1,9 +1,19 @@
...
+  - export PATH=$PATH:$PWD/kratos
+
 install:
   - pip install -r dev_requirements.txt
   - python setup.py develop
 
 script:
-  - flake8 --exclude=CFD_example.gid,3balls.gid --ignore=F401,F403 .
\ No newline at end of file
+  - flake8 --exclude=use_cases,utilities,kratos,ProjectParameters.py,DEM_explicit_solver_var.py --ignore=F403 .
```
Comment (itziakos) [inline .travis.yml]: you should add the specifics into a configuration file, see http://flake8.readthedocs.org/en/latest/config.html#global

File: README.rst [hunk 18]
```diff
@@ -15,11 +15,56 @@ simphony-kratos is hosted on github: https://github.com/simphony/simphony-kratos
...
 
-- somphony-common > 0.1.1
+- `simphony-common`_ == 0.1.3
+
+.. _simphony-common: https://github.com/simphony/simphony-common
 
 Installation
 ------------
 
+Installation is based on setuptools and requeires Python 2.7::
```
Comment (nathanfranklin) [inline README.rst]: requeires -> requires

File: README.rst [hunk 19]
```diff
@@ -15,11 +15,56 @@ simphony-kratos is hosted on github: https://github.com/simphony/simphony-kratos
...
 Kratos Installation
--------------------
+~~~~~~~~~~~~~~~~~~~
+
+Both Kratos-CFD and DEMPack wrappers use KratosMultiphysics. A suitable version to run with these wrappers can be downloaded here:
+
+- `KratosMultiphysics`_
+
+.. _KratosMultiphysics: https://web.cimne.upc.edu/users/croig/data/kratos-simphony.tgz
+
```
Comment (nathanfranklin) [inline README.rst:43]: Maybe additonal information could be added about downloadable version of Kratos:

This version is prepared for Ubuntu-12-04-LTS 64-bit.  See http://kratos-wiki.cimne.upc.edu/index.php/Download for additional information on getting Kratos for a specific system.

Comment (roigcarlo) [inline README.rst:43]: Sure. I upload it in the next push

**Comment by roigcarlo** [comment]
@nathanfranklin Apparently tests are passing now. I believe I have corrected all the suggestions, but please let me double check it again.

After reading @itziakos suggestion both CFD and DEM are under the same import, which I would like to change in a near future. Anyway, this should be compliant with the API so #15 ( I also changed the test so they use `from simphony.engine import kratos`) and #11 should be fixed at this point.

**Comment by itziakos** [comment]
> After reading @itziakos suggestion both CFD and DEM are under the same import, which I would like to change in a near future. Anyway, this should be compliant with the API so #15 ( I also changed the test so they use from simphony.engine import kratos) and #11 should be fixed at this point.

@roigcarlo, I did not comment on the multiple entry points lines of the setup.py. There is nothing wrong with that. My comments were specifically about the name of the package (wrappers -> simphony_kratos or something else) and the fact that the setup.py was doing unnecessary things (the version.py file is only related to the package not independent modules).

I feel that there is a confusion and probably we are using different terms. Each simphony plugin is a single python package that contains multiple subpackages and submodules. The name of the kratos plugin was `wrappers` which was a very generic name. The simphony plugin packages can register simphony engine entry points which they discribe in the setup.py. There is no limitation or problem with registering multiple engines from a single package. In sort the name of the kratos package and the name or number of engine entry points is independent.

**Comment by roigcarlo** [comment]
@itziakos Thx for the explanation, its very helpful. It is the first time I touched the entire package-module-plugin thing and I am a little confused, but I believe I see your point now. Basically if i understand it correctly we can have:

```
simkratos            package 
→ kratosCFD          plugin1
→ DEMPack            plugin2
```

I think that, for the initial release, it is ok to leave things like they are. I just wanted ( and still want ) to use the multiple entry points to make clear that, even sharing part of the code they are completely separate things. Also I think that it could be interesting to extend the package with other Kratos solvers since they are already there and should be relativelly easy to implement.

Thx again for taking time to explain me the difference.

P.S: Should we merge this to close #11 #14 and #15?

**Comment by itziakos** [comment]
> P.S: Should we merge this to close #11 #14 and #15?

I mainly checked the setup.py and travis.yml and did not spent much to time looking at other parts of this large PR. I think that @nathanfranklin is the most appropriate person to answer this since he have been working on the integration and following the changes.

**Comment by itziakos** [comment]
> Also I think that it could be interesting to extend the package with other Kratos solvers since they are already there and should be relativelly easy to implement.

Yes, The fact that `kratos` already has a functioning python layer makes it easier.

**Comment by nathanfranklin** [comment]
> >  P.S: Should we merge this to close #11 #14 and #15?
> 
> I mainly checked the setup.py and travis.yml and did not spent much to time looking at other parts of this large PR. I think that @nathanfranklin is the most appropriate person to answer this since he have been working on the integration and following the changes.

Looks good to me.  Thanks.

I will integrate on monday and double check if there are any issues.  If all is well, then you make a new release 0.1.1 of Kratos.

**Comment by nathanfranklin** [comment]
@roigcarlo ,  all looks good on integration.  just needed to add the mdpa (have commit for you to review at #16).  

After this #9 is merged, can you immediately make a new release (0.1.1 of simphony-kratos). Thanks.

**Comment by roigcarlo** [comment]
@nathanfranklin Perfect. I Merged #16. As soon as it passes I will merge this and make a release

**Comment by nathanfranklin** [comment]
@roigcarlo thanks.
